File: jar/src/main/java/jline/lib/butools/APHFrom3Moments.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.lib.butools;

import org.apache.commons.math3.complex.Complex;

import jline.GlobalConstants;
import jline.lang.processes.APH;
import jline.util.Maths;
import jline.util.matrix.Matrix;

public final class APHFrom3Moments {
    private APHFrom3Moments() {}

    public static APH APHFrom3Moments(double[] moms) {
        return APHFrom3Moments(moms, 100, 1e-14);
    }

    public static APH APHFrom3Moments(double[] moms, int maxSize) {
        return APHFrom3Moments(moms, maxSize, 1e-14);
    }

    public static APH APHFrom3Moments(double[] moms, int maxSize, double prec) {
        if (maxSize <= 0) {
            maxSize = 100;
        }
        if (prec <= 0) {
            prec = 1e-14;
        }

        double m1 = moms[0];
        double m2 = moms[1];
        double m3 = moms[2];

        // Detect number of phases needed
        int n = 2;
        while (n < maxSize && (APH2ndMomentLowerBound.APH2ndMomentLowerBound(m1, n) > m2
                || APH3rdMomentLowerBound.APH3rdMomentLowerBound(m1, m2, n) >= m3
                || APH3rdMomentUpperBound.APH3rdMomentUpperBound(m1, m2, n) <= m3)) {
            n++;
        }

        if (APH2ndMomentLowerBound.APH2ndMomentLowerBound(m1, n) > m2) {
            m2 = APH2ndMomentLowerBound.APH2ndMomentLowerBound(m1, n);
        }

        if (APH3rdMomentLowerBound.APH3rdMomentLowerBound(m1, m2, n) > m3) {
            m3 = APH3rdMomentLowerBound.APH3rdMomentLowerBound(m1, m2, n);
        }

        if (APH3rdMomentUpperBound.APH3rdMomentUpperBound(m1, m2, n) < m3) {
            m3 = APH3rdMomentUpperBound.APH3rdMomentUpperBound(m1, m2, n);
        }

        // Compute normalized moments
        double[] nmoms = NormMomsFromMoms.NormMomsFromMoms(new double[]{m1, m2, m3});
        double n1 = nmoms[0];
        double n2 = nmoms[1];
        double n3 = nmoms[2];

        if (n2 > 2 || n3 < 2 * n2 - 1) {
            Complex nComplex = new Complex((double) n, 0.0);
            Complex n2Complex = new Complex(n2, 0.0);
            Complex n3Complex = new Complex(n3, 0.0);
            Complex four = new Complex(4.0, 0.0);
            Complex two = new Complex(2.0, 0.0);
            Complex three = new Complex(3.0, 0.0);
            Complex twelve = new Complex(12.0, 0.0);
            Complex sixteen = new Complex(16.0, 0.0);
            Complex fifteen = new Complex(15.0, 0.0);
            Complex eight = new Complex(8.0, 0.0);
            Complex one = Complex.ONE;

            Complex numeratorB = two.multiply(four.subtract(nComplex.multiply(three.multiply(n2Complex).subtract(four))));
            Complex denominatorB = n2Complex.multiply(four.add(nComplex).subtract(nComplex.multiply(n3Complex)))
                    .add((nComplex.multiply(n2Complex)).sqrt()
                            .multiply((twelve.multiply(n2Complex.multiply(n2Complex)).multiply(nComplex.add(one))
                                    .add(sixteen.multiply(n3Complex).multiply(nComplex.add(one)))
                                    .add(n2Complex.multiply(nComplex.multiply(n3Complex.subtract(fifteen)).multiply(n3Complex.add(one))
                                            .subtract(eight.multiply(n3Complex.add(new Complex(3.0, 0.0)))))).sqrt())));

            Complex b = numeratorB.divide(denominatorB);

            Complex numeratorA = (b.multiply(n2Complex).subtract(two)).multiply(nComplex.subtract(one)).multiply(b);
            Complex denominatorA = b.subtract(one).multiply(nComplex);
            Complex a = numeratorA.divide(denominatorA);

            Complex p = (b.subtract(one)).divide(a);

            Complex lambda = (p.multiply(a).add(one)).divide(n1);

            Complex mu = (nComplex.subtract(one)).multiply(lambda).divide(a);

            // Construct representation
            double[] alpha = new double[n];
            alpha[0] = p.getReal();
            alpha[n - 1] = 1.0 - p.getReal();

            Matrix A = new Matrix(n, n);
            A.set(n - 1, n - 1, -lambda.getReal());
            for (int i = 0; i < n - 1; i++) {
                A.set(i, i, -mu.getReal());
                A.set(i, i + 1, mu.getReal());
            }
            return new APH(new Matrix(alpha).transpose(), A);
        } else {
            double c4 = n2 * (3.0 * n2 - 2.0 * n3) * Math.pow((double) (n - 1), 2.0);
            double c3 = 2.0 * n2 * (n3 - 3.0) * Math.pow((double) (n - 1), 2.0);
            double c2 = 6.0 * (n - 1.0) * (n - n2);
            double c1 = 4.0 * n * (2.0 - n);
            double c0 = n * (n - 2.0);

            double[] coefficients = new double[]{c0, c1, c2, c3, c4};
            Complex[] fs = Maths.roots(coefficients);

            for (Complex f : fs) {
                if (f.isNaN() || f.isInfinite()) {
                    continue;
                }

                Complex temp1 = f.multiply(f).multiply(n2).subtract(f.multiply(2.0)).add(2.0);
                Complex temp2 = temp1.multiply(n - 1);
                Complex temp3 = temp2.subtract((double) n);

                if (temp3.abs() < prec) {
                    continue;
                }

                Complex a = f.subtract(1.0).multiply(2.0).multiply(n - 1).divide(temp3);
                Complex p = f.subtract(1.0).multiply(a);
                Complex lambda = a.add(p).divide(n1);
                Complex mu = new Complex((double) (n - 1)).divide(n1 - p.divide(lambda).getReal());

                if (Math.abs(p.getImaginary()) < GlobalConstants.Zero) {
                    if (Math.abs(lambda.getImaginary()) < GlobalConstants.Zero) {
                        if (Math.abs(mu.getImaginary()) < GlobalConstants.Zero) {
                            if (p.getReal() >= 0 && p.getReal() <= 1 && lambda.getReal() > 0 && mu.getReal() > 0) {
                                double[] alpha = new double[n];
                                alpha[0] = p.getReal();
                                alpha[1] = 1.0 - p.getReal();

                                Matrix A = new Matrix(n, n);
                                A.set(0, 0, -lambda.getReal());
                                A.set(0, 1, lambda.getReal());
                                for (int j = 1; j < n; j++) {
                                    A.set(j, j, -mu.getReal());
                                    if (j < n - 1) {
                                        A.set(j, j + 1, mu.getReal());
                                    }
                                }
                                return new APH(new Matrix(alpha).transpose(), A);
                            }
                        }
                    }
                }
            }
        }

        throw new IllegalArgumentException("No APH found for the given 3 moments!");
    }
}

File: jar/src/main/kotlin/jline/lib/butools/APHFrom3Moments.kt
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.lib.butools

import jline.GlobalConstants
import jline.lang.processes.APH
import jline.util.Maths
import jline.util.matrix.Matrix
import org.apache.commons.math3.complex.Complex
import kotlin.math.abs
import kotlin.math.pow


fun APHFrom3Moments(moms: DoubleArray, maxSize: Int = 100, prec: Double = 1e-14): APH {
    var maxSize = maxSize
    var prec = prec
    if (maxSize <= 0) {
        maxSize = 100
    }
    if (prec <= 0) {
        prec = 1e-14
    }

    val m1 = moms[0]
    var m2 = moms[1]
    var m3 = moms[2]

    // Detect number of phases needed
    var n = 2
    while (n < maxSize && (APH2ndMomentLowerBound(m1, n) > m2 || APH3rdMomentLowerBound(m1,
            m2,
            n) >= m3 || APH3rdMomentUpperBound(m1, m2, n) <= m3)) {
        n++
    }

    // If PH is too large, adjust moment to bounds
    if (APH2ndMomentLowerBound(m1, n) > m2) {
        m2 = APH2ndMomentLowerBound(m1, n)
    }

    if (APH3rdMomentLowerBound(m1, m2, n) > m3) {
        m3 = APH3rdMomentLowerBound(m1, m2, n)
    }

    if (APH3rdMomentUpperBound(m1, m2, n) < m3) {
        m3 = APH3rdMomentUpperBound(m1, m2, n)
    }

    // Compute normalized moments
    val nmoms = NormMomsFromMoms(doubleArrayOf(m1, m2, m3))
    val n1 = nmoms[0]
    val n2 = nmoms[1]
    val n3 = nmoms[2]


    if (n2 > 2 || n3 < 2 * n2 - 1) {
        val nComplex = Complex(n.toDouble(), 0.0)
        val n2Complex = Complex(n2, 0.0)
        val n3Complex = Complex(n3, 0.0)
        val four = Complex(4.0, 0.0)
        val two = Complex(2.0, 0.0)
        val three = Complex(3.0, 0.0)
        val twelve = Complex(12.0, 0.0)
        val sixteen = Complex(16.0, 0.0)
        val fifteen = Complex(15.0, 0.0)
        val eight = Complex(8.0, 0.0)
        Complex(-1.0, 0.0)
        val one = Complex.ONE

        val numeratorB = two.multiply(four.subtract(nComplex.multiply(three.multiply(n2Complex).subtract(four))))
        val denominatorB = n2Complex.multiply(four.add(nComplex).subtract(nComplex.multiply(n3Complex)))
            .add((nComplex.multiply(n2Complex)).sqrt()
                .multiply((twelve.multiply(n2Complex.multiply(n2Complex)).multiply(nComplex.add(one))
                    .add(sixteen.multiply(n3Complex).multiply(nComplex.add(one)))
                    .add(n2Complex.multiply(nComplex.multiply(n3Complex.subtract(fifteen)).multiply(n3Complex.add(one))
                        .subtract(eight.multiply(n3Complex.add(Complex(3.0, 0.0)))))).sqrt())))

        val b = numeratorB.divide(denominatorB)

        val numeratorA = (b.multiply(n2Complex).subtract(two)).multiply(nComplex.subtract(one)).multiply(b)
        val denominatorA = b.subtract(one).multiply(nComplex)
        val a = numeratorA.divide(denominatorA)

        val p = (b.subtract(one)).divide(a)

        val lambda = (p.multiply(a).add(one)).divide(n1)

        val mu = (nComplex.subtract(one)).multiply(lambda).divide(a)

        // Construct representation
        val alpha = DoubleArray(n)
        alpha[0] = p.real
        alpha[n - 1] = 1.0 - p.real

        val A = Matrix(n, n)
        A[n - 1, n - 1] = -lambda.real
        for (i in 0..<n - 1) {
            A[i, i] = -mu.real
            A[i, i + 1] = mu.real
        }
        return APH(Matrix(alpha).transpose(), A)
    } else {
        val c4 = n2 * (3.0 * n2 - 2.0 * n3) * (n - 1).toDouble().pow(2.0)
        val c3 = 2.0 * n2 * (n3 - 3.0) * (n - 1).toDouble().pow(2.0)
        val c2 = 6.0 * (n - 1.0) * (n - n2)
        val c1 = 4.0 * n * (2.0 - n)
        val c0 = n * (n - 2.0)

        val coefficients = doubleArrayOf(c0, c1, c2, c3, c4)
        val fs = Maths.roots(coefficients)

        for (f in fs) {
            if (f.isNaN || f.isInfinite) {
                continue
            }

            val temp1 = f.multiply(f).multiply(n2).subtract(f.multiply(2.0)).add(2.0)
            val temp2 = temp1.multiply(n - 1)
            val temp3 = temp2.subtract(n.toDouble())

            if (temp3.abs() < prec) {
                continue
            }

            val a = f.subtract(1.0).multiply(2.0).multiply(n - 1).divide(temp3)
            val p = f.subtract(1.0).multiply(a)
            val lambda = a.add(p).divide(n1)
            val mu = Complex((n - 1).toDouble()).divide(n1 - p.divide(lambda).real)

            if (abs(p.imaginary) < GlobalConstants.Zero) if (abs(lambda.imaginary) < GlobalConstants.Zero) {
                if (abs(mu.imaginary) < GlobalConstants.Zero) {
                    if (p.real >= 0 && p.real <= 1 && lambda.real > 0 && mu.real > 0) {
                        val alpha = DoubleArray(n)
                        alpha[0] = p.real
                        alpha[1] = 1.0 - p.real

                        val A = Matrix(n, n)
                        A[0, 0] = -lambda.real
                        A[0, 1] = lambda.real
                        for (j in 1..<n) {
                            A[j, j] = -mu.real
                            if (j < n - 1) {
                                A[j, j + 1] = mu.real
                            }
                        }
                        // Return or further process alpha and A as needed
                        return APH(Matrix(alpha).transpose(), A)
                    }
                }
            }
        }
    }

    throw IllegalArgumentException("No APH found for the given 3 moments!")
}

